File: evolve/src/animal/animal.rs
use std::num::SignedInt;
use std::collections::HashMap;

pub struct Animal {
    pub x:      i32,
    pub y:      i32,
    pub energy: i32,
    pub dir:    i32,
    pub genes:  Vec<i32>,
    pub alive:  bool,
}

impl Animal {

    pub fn new(x: i32, y: i32, e: i32, d: i32, g: Vec<i32>, a: bool) -> Animal {
        Animal { x: x, y: y, energy: e, dir: d, genes: g, alive: a }
    }

    pub fn show(&self) {
        println!("x: {}, y: {}\nenergy: {}\ndir: {}\nalive: {}\ngenes: {:?}",
                 self.x, self.y, self.energy, self.dir, self.alive, self.genes)
    }

    pub fn ani_move(&mut self, width: i32, height: i32) {

        self.x = ( self.x + (if self.dir >= 2 && self.dir < 5 { 1 }
                else if self.dir == 1 || self.dir == 5 { 0 }
                else { -1 }) + width) % width;

        self.y = (self.y + (if self.dir >= 0 && self.dir < 3 { -1 }
                else if self.dir >= 4 || self.dir < 7 { 1 }
                else { 0 }) + height) % height;

        self.energy -= 1;
    }

    pub fn turn(&mut self) {

        // this needs work - add in something similar to what he has.
        if self.genes[((self.x + self.y) % 8) as usize] > 5 {
            self.dir += 1;
        } else {
            self.dir -= 1;
        }

        self.dir = SignedInt::abs(self.dir % 8);
    }
    
    pub fn eat(&mut self, plants: &mut HashMap<(i32, i32), bool>, plant_energy: i32) {

        let pos = (self.x, self.y);

        if plants.contains_key(&pos) {
            self.energy += plant_energy;
            plants.remove(&pos);
        }
    }

    pub fn reproduce(&mut self, rep_energy: i32) -> bool {

        if self.energy >= rep_energy {
            self.energy /= 2;
            return true
        }
        false
    }

    pub fn is_alive(&mut self) {

        if self.energy < 1 { self.alive = false };
    }
}

File: some_c/Evolve/animal.c
#include "animal.h"

void
show_animal(animal *a)
{
    int i;
    
    printf("x: %d, y: %d\nenergy: %d\ndir: %d\nliving: %s\nGenes: ",
        a->x, a->y, a->energy, a->dir, (a->alive == 1) ? "true" : "false");
    
    for (i = 0; i < 8; i++) {
        printf("%d ", a->genes[i]);
    }
    printf("\n");
}

void
move_animal(animal *a, int w, int h)
{
    a->x = (a->x + ((a->dir >= 2 && a->dir <  5) ? 1 :
                    (a->dir == 1 || a->dir == 5) ? 0 : 
                                                  -1) + w) % w;
                                          
    a->y = (a->y + ((a->dir >= 0 && a->dir < 3) ? -1 :
                    (a->dir >= 4 || a->dir < 7) ?  1 : 
                                                   0) + h) % h;
    
    a->energy -= 1;
}

void
turn_animal(animal *a)
{
   a->dir += (a->genes[(a->x + a->y) % 8] > 5) ? 1 : -1; 
   a->dir = abs(a->dir);
}

//write this with a plain array
void
eat_animal(animal *a, plant *p[])
{
    int i;

    for (i = 0; i < sizeof(p); i++) {
	 if (a->x == p[i]->x && a->y == p[i]->y) p[i] = 0;
    }	 
}

bool
reproduce_animal(animal *a, int rep_energy)
{
    if (a->energy > rep_energy) {
        a->energy /= 2;
        return true;
    }
    return false;
}

void
still_alive_animal(animal *a)
{
    if (a->energy < 1) a->alive = false;
}

animal
copy_animal(animal *a)
{
    animal b;
    
    b.x = a->x;
    b.y = a->y;
    b.energy = a->energy;
    b.dir = (a->dir + 1) % 8;
    // b.genes = copy_genes(&a->genes);
    gen_genes(&b);
    b.alive = true;
    mutate_gene(&b);
    
    return b;
}

void
init_animal(animal *a)
{
    a->x = WIDTH / 2;
    a->y = HEIGHT / 2;
    a->energy = 1000;
    a->dir = 0;
    a->alive = true;
    gen_genes(a);
}